src/extlibs/hrvo/new_simulator.h

```diff
@@ -0,0 +1,119 @@
+#pragma once
+
+#include <fstream>
+#include <limits>
+#include <vector>
+
+#include "extlibs/hrvo/agent.h"
+#include "extlibs/hrvo/kd_tree.h"
+#include "proto/tbots_software_msgs.pb.h"
+#include "proto/visualization.pb.h"
+#include "software/geom/vector.h"
+#include "software/world/world.h"
+#include "software/geom/algorithms/intersection.h"
+#include "software/geom/algorithms/nearest_neighbor_search.hpp"
+
+class MotionPlanningSimulator {
+public:
+    /**
+     * Constructor
+     * @param time_step The time step between each step of the simulator
+     * @param robot_constants The robot constants to be used for all Agents representing a
+     * robot
+     * @param friendly_team_colour The colour of the friendly team
+     */
+    explicit MotionPlanningSimulator(float time_step, const RobotConstants_t &robot_constants,
+                                     const TeamColour friendly_team_colour);
+
+    /**
+     * Reset all agents to match the state of the given world.
+     * Friendly robots will use the Hybrid Reciprocal algorithm to traverse.
+     * Enemy robots will go directly towards their goal without trying to avoid any
+     * obstacles
+     *
+     * @param world The world which the simulation should be based upon
+     */
+    void updateWorld(const World &world);
+
+    /**
+     * Reset all friendly agents goal points to match the path of the given primitive set
+     *
+     * @param new_primitive_set
+     */
+    void updatePrimitiveSet(const TbotsProto::PrimitiveSet &new_primitive_set);
+
+    void doStep(double time_step);
+
+    std::optional<std::shared_ptr<HRVOAgent>> getFriendlyAgentFromRobotId(
+            unsigned int robot_id) const;
+
+
+private:
+
+    /**
+     *      Adds a new Hybrid Reciprocal Agent to the simulation based on Robot.
+     *
+     * @param robot    The robot which this agent should be based on
+     * @param type     Whether this robot is FRIENDLY or ENEMY
+     *
+     * @return    The index of the agent.
+     */
+    std::size_t addHRVORobotAgent(const Robot &robot, TeamSide type);
+
+    /**
+     *      Adds a new Linear Velocity Agent to the simulation based on Robot.
+     *
+     * @param robot       	The robot which this agent should be based on
+     * @param destination 	Destination for this robot
+     * @param type 		Whether this robot is FRIENDLY or ENEMY
+     * @return    The index of the agent.
+     */
+    std::size_t addLinearVelocityRobotAgent(const Robot &robot, const Vector &destination,
+                                            TeamSide type);
+
+    // PrimitiveSet which includes the path which each friendly robot should take
+    TbotsProto::PrimitiveSet primitive_set;
+
+    // Latest World which the simulator has received
+    std::optional<World> world;
+
+    // The robot constants which all agents will use
+    RobotConstants_t robot_constants;
```

Comment:
`robot_constants` should probably not be needed in the simulator class. Ideally it would only exist in the `Agent` class